collections/src/main/scala/strawman/collection/immutable/ChampHashMap.scala

```diff
@@ -0,0 +1,601 @@
+
+  override final def contains(key: K): Boolean = rootNode.containsKey(key, computeHash(key), 0)
+
+  def get(key: K): Option[V] = rootNode.get(key, computeHash(key), 0)
+
+  def updated[V1 >: V](key: K, value: V1): ChampHashMap[K, V1] = {
+    val effect = MapEffect[K, V1]()
+    val keyHash = computeHash(key)
+    val newRootNode = rootNode.updated(key, value, keyHash, 0, effect)
+
+    if (effect.isModified) {
+      if (effect.hasReplacedValue) {
+        ChampHashMap(newRootNode, cachedJavaHashCode, cachedSize)
+      } else {
+        ChampHashMap(newRootNode, cachedJavaHashCode + keyHash, cachedSize + 1)
+      }
+    }
+    else this
+  }
+
+  def remove(key: K): ChampHashMap[K, V] = {
+    val effect = MapEffect[K, V]()
+    val keyHash = computeHash(key)
+    val newRootNode = rootNode.removed(key, keyHash, 0, effect)
+
+    if (effect.isModified)
+      ChampHashMap(newRootNode, cachedJavaHashCode - keyHash, cachedSize - 1)
+    else this
+  }
+
+  def empty: ChampHashMap[K, V] = ChampHashMap.empty
+
+  override def equals(that: Any): Boolean =
+    that match {
+      case map: ChampHashMap[K, V] =>
+        (this eq map) ||
+          (this.cachedSize == map.cachedSize) &&
+            (this.cachedJavaHashCode == map.cachedJavaHashCode) &&
+            (this.rootNode == map.rootNode)
+      case _ => super.equals(that)
+    }
+
+}
+
+private[immutable] object MapNode {
+
+  private final val EmptyMapNode = new BitmapIndexedMapNode(0, 0, Array.empty)
+
+  def empty[K, V]: MapNode[K, V] =
+    EmptyMapNode.asInstanceOf[MapNode[K, V]]
+
+  final val TupleLength = 2
+
+}
+
+@SerialVersionUID(2L)
+private[immutable] sealed abstract class MapNode[K, +V] extends Node[MapNode[K, V @uV]] with Serializable {
+
+  def get(key: K, hash: Int, shift: Int): Option[V]
+
+  def containsKey(key: K, hash: Int, shift: Int): Boolean =
+    this.get(key, hash, shift).isDefined
+
+  def contains[V1 >: V](key: K, value: V1, hash: Int, shift: Int): Boolean =
+    this.get(key, hash, shift).contains(value)
+
+  def updated[V1 >: V](key: K, value: V1, hash: Int, shift: Int, effect: MapEffect[K, V1]): MapNode[K, V1]
+
+  def removed[V1 >: V](key: K, hash: Int, shift: Int, effect: MapEffect[K, Any]): MapNode[K, V1]
+
+  def hasNodes: Boolean
+
+  def nodeArity: Int
+
+  def getNode(index: Int): MapNode[K, V]
+
+  def hasPayload: Boolean
+
+  def payloadArity: Int
+
+  def getPayload(index: Int): (K, V)
+
+  def sizePredicate: Int
+
+}
+
+@SerialVersionUID(2L)
+private[this] final class BitmapIndexedMapNode[K, +V](val dataMap: Int, val nodeMap: Int, val content: Array[Any]) extends MapNode[K, V] {
+
+  import Node._
+  import MapNode._
+
+  assert(TupleLength * payloadArity + nodeArity == content.length)
+  assert(Range(0, TupleLength * payloadArity).forall(i => !content(i).isInstanceOf[MapNode[_, _]]))
+  assert(Range(TupleLength * payloadArity, content.length).forall(i => content(i).isInstanceOf[MapNode[_, _]]))
```

Comment:
These assertions will be evaluated by user code unless they use a specific compiler option. Is it possible to move them to tests instead?